src/enzo/Grid_ComputeDomainBoundaryMassFlux.C

```diff
@@ -0,0 +1,140 @@
+/**********************************************************************
+/
+/  GRID CLASS (COMPUTE DOMAIN BOUNDARY MASS FLUX)
+/
+/  written by: Andrew Emerick
+/  date:       January, 2017
+/  modified1:
+/
+/  PURPOSE:
+/    For density field and all species fields, compute
+/    mass flux on domain boundary to count mass that
+/    has left the domain. Stores this in global counter
+/
+/  RETURNS:
+/     SUCCESS or FAIL
+***********************************************************************/
+
+#include <stdio.h>
+#include <math.h>
+#include "ErrorExceptions.h"
+#include "macros_and_parameters.h"
+#include "typedefs.h"
+#include "global_data.h"
+#include "Fluxes.h"
+#include "GridList.h"
+#include "ExternalBoundary.h"
+#include "Grid.h"
+
+int GetUnits(float *DensityUnits, float *LengthUnits,
+             float *TemperatureUnits, float *TimeUnits,
+             float *VelocityUnits, FLOAT Time);
+
+
+int grid::ComputeDomainBoundaryMassFlux(float *allgrid_BoundaryMassFluxContainer,
+                                        TopGridData *MetaData)
+{
+
+  /* Return if this doesn't concern us */
+
+  if (ProcessorNumber != MyProcessorNumber)
+    return SUCCESS;
+
+  PrepareBoundaryMassFluxFieldNumbers(); //  this should be put somewhere else
+
+  /* get units */
+  const float msolar = 1.989e33;
+  float DensityUnits, LengthUnits, TemperatureUnits, TimeUnits, VelocityUnits, MassUnits;
+  if (GetUnits(&DensityUnits, &LengthUnits, &TemperatureUnits,
+               &TimeUnits, &VelocityUnits, this->Time) == FAIL){
+      ENZO_FAIL("Error in GetUnits");
+  }
+  MassUnits   = DensityUnits*LengthUnits*LengthUnits*LengthUnits; // mass unit
+
+  /* Is this grid on the edge of the domain */
+  int GridOffsetLeft[MAX_DIMENSION], GridOffsetRight[MAX_DIMENSION];
+  for (int dim = 0; dim < MAX_DIMENSION; dim++){
+    if (dim < this->GridRank){
+      GridOffsetLeft[dim] = nint((this->GridLeftEdge[dim] - DomainLeftEdge[dim])/
+                                  this->CellWidth[dim][0]);
+      GridOffsetRight[dim] = nint((this->GridRightEdge[dim] - DomainRightEdge[dim])/
+                                  this->CellWidth[dim][0]);
+    } else{
+      GridOffsetLeft[dim] = 0; GridOffsetRight[dim] = 0;
+    }
+  }
+
+  int NumberOfBoundaryMassFields   = 1; // density always
+
+  if (MultiSpecies > 0)
+    NumberOfBoundaryMassFields += 5; // HI, HII, HeI, HeII, HeIII
+
+  if (MultiSpecies > 1)
+    NumberOfBoundaryMassFields += 3; // H2, H2I, HM
+
+  if (MultiSpecies > 2)
+    NumberOfBoundaryMassFields += 3; // D, D+, HD
+
+  if (MultiMetals > 0)
```

Comment:
`MultiMetals` isn't always set to be >0 when metal fields are being used and multiple metal fields might be used. You could use `grid::IdentifyColourFields` to count the metal fields when the simulation starts and then store this number somewhere.